### PR TITLE
Fix usage of infrastructure labels/annotations on gateway

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/50983, I think

Basically, we have logic to use `infra.Labels || meta.Labels`. We modify
meta.Labels.

This is wrong, we should modify AFTER we pick which set of labels to
use.

Also, we need to deepcopy anything we mutate since its accessing a
shared object cache

### CODE REVIEW

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 1]
```diff
@@ -450,6 +399,58 @@ func (d *DeploymentController) configureIstioGateway(log *istiolog.Scope, gw gat
...
+	}
+
+	// Default the network label for waypoints if not explicitly set in gateway's labels
+	network := d.injectConfig().Values.Struct().GetGlobal().GetNetwork()
+	if _, ok := gw.GetLabels()[label.TopologyNetwork.Name]; !ok && network != "" && isWaypointGateway {
+		input.InfrastructureLabels[label.TopologyNetwork.Name] = d.injectConfig().Values.Struct().GetGlobal().GetNetwork()
+	}
+}
+
+func extractInfrastructureMetadata(gwInfra *gatewayv1.GatewayInfrastructure, isLabel bool, gw gateway.Gateway) map[string]string {
```
Comment (keithmattix) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:426]: nit: don't love the bool arg; would prefer clearly named functions using a common base, but not a big deal

**Review by keithmattix — APPROVED** [review]
Nice find! LGTM just adding a hold in case you wanted to apply the nit; remove at your convenience

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:430]: :+1: done

**Comment by howardjohn** [comment]
(Removed hold as I addressed the comment)

**Comment by howardjohn** [comment]
the bug is in the interaction of 2 new features, not really a signal of
overall stability

**Pull request merged**